n2v/methods/oucarter.py

```diff
@@ -0,0 +1,351 @@
+        (i.e. the 2dn and 3rd term in eqn. (17) in [1]):
+        """
+
+        if grid_info is None:
+            tauLmP_rho = np.zeros(self.Vpot.grid().npoints())
+            nblocks = self.Vpot.nblocks()
+
+            points_func = self.Vpot.properties()[0]
+            blocks = None
+
+        else:
+            blocks, npoints, points_func = grid_info
+            tauLmP_rho = np.zeros(npoints)
+            nblocks = len(blocks)
+
+        points_func.set_deriv(2)
+
+        iw = 0
+        for l_block in range(nblocks):
+            # Obtain general grid information
+            if blocks is None:
+                l_grid = self.Vpot.get_block(l_block)
+            else:
+                l_grid = blocks[l_block]
+            l_npoints = l_grid.npoints()
+
+            points_func.compute_points(l_grid)
+            l_lpos = np.array(l_grid.functions_local_to_global())
+            if len(l_lpos) == 0:
+                iw += l_npoints
+                continue
+            l_phi = np.array(points_func.basis_values()["PHI"])[:l_npoints, :l_lpos.shape[0]]
+            l_phi_x = np.array(points_func.basis_values()["PHI_X"])[:l_npoints, :l_lpos.shape[0]]
+            l_phi_y = np.array(points_func.basis_values()["PHI_Y"])[:l_npoints, :l_lpos.shape[0]]
+            l_phi_z = np.array(points_func.basis_values()["PHI_Z"])[:l_npoints, :l_lpos.shape[0]]
+            l_phi_xx = np.array(points_func.basis_values()["PHI_XX"])[:l_npoints, :l_lpos.shape[0]]
+            l_phi_yy = np.array(points_func.basis_values()["PHI_YY"])[:l_npoints, :l_lpos.shape[0]]
+            l_phi_zz = np.array(points_func.basis_values()["PHI_ZZ"])[:l_npoints, :l_lpos.shape[0]]
+
+            lD = D[(l_lpos[:, None], l_lpos)]
+            # lC = C[l_lpos, :]
+
+            rho = contract('pm,mn,pn->p', l_phi, lD, l_phi)
+            rho_inv = 1/rho
+
+            # Calculate the second term 0.25*\nabla^2\rho
+            laplace_rho_temp = contract('ab,pa,pb->p', lD, l_phi, l_phi_xx + l_phi_yy + l_phi_zz)
+            # laplace_rho_temp += contract('pm, mn, pn->p', l_phi_x,lD, l_phi_x)
+            # laplace_rho_temp += contract('pm, mn, pn->p', l_phi_y,lD, l_phi_y)
+            # laplace_rho_temp += contract('pm, mn, pn->p', l_phi_z,lD, l_phi_z)
+            laplace_rho_temp += np.sum((l_phi_x @ lD) * l_phi_x, axis=1)
+            laplace_rho_temp += np.sum((l_phi_y @ lD) * l_phi_y, axis=1)
+            laplace_rho_temp += np.sum((l_phi_z @ lD) * l_phi_z, axis=1)
+
+            laplace_rho_temp *= 0.25 * 2
+
+            # Calculate the third term |nabla rho|^2 / 8
+            tauW_temp = contract('pm, mn, pn->p', l_phi, lD, l_phi_x) ** 2
+            tauW_temp += contract('pm, mn, pn->p', l_phi, lD, l_phi_y) ** 2
+            tauW_temp += contract('pm, mn, pn->p', l_phi, lD, l_phi_z) ** 2
+            tauW_temp *= rho_inv * 0.125 * 4
+
+            tauLmP_rho[iw: iw + l_npoints] = (-laplace_rho_temp + tauW_temp) * rho_inv
```

Comment:
I see that in equation (17) only the second term is divided by rho but here both terms are divided by rho. But here we multiply by the inverse even after line 81. Am I missing something?

Reply:
So precisely $\frac{\tau_L}{\rho}$ is need for the calculation instead of $\tau_L$.